sstables/sstable_set.cc

```diff
@@ -0,0 +1,526 @@
+            throw;
+        }
+    } catch (...) {
+        _current_sstable_versions.erase(sst);
+        throw;
+    }
+}
+
+void
+sstable_set::erase(shared_sstable sst, bool capturing) {
+    if (!_current_sstable_versions.contains(sst)) {
+        return;
+    }
+    if (capturing && _version_references[_current_version] != nullptr) {
+        forward_version();
+    }
+
+    auto vsst = _current_sstable_versions.at(sst);
+    if (_added_at[_current_version].contains(vsst)) {
+        // the sstable was erased before being read by anyone
+        _added_at[_current_version].erase(vsst);
+        remove(vsst);
+    } else {
+        _erased_at[_current_version].insert(vsst);
+    }
+
+    vsst->version_erased = _current_version;
+    _current_sstable_versions.erase(sst);
+    _current_all->_size--;
+}
+
+sstable_set::~sstable_set() = default;
+
+version_reference::version_reference(lw_shared_ptr<sstable_set> versioned_ssts, uint64_t version)
+    : _versioned_ssts(std::move(versioned_ssts)), _version(version) {
+}
+
+version_reference::~version_reference() {
+    _versioned_ssts->remove_reference(_version);
+}
+
+uint64_t version_reference::get() const {
+    return _version;
+}
+
+sstable_set::incremental_selector::incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s, uint64_t version)
+    : _impl(std::move(impl))
+    , _cmp(s)
+    , _version(version) {
+}
+
+sstable_set::incremental_selector::~incremental_selector() = default;
+
+sstable_set::incremental_selector::incremental_selector(sstable_set::incremental_selector&&) noexcept = default;
+
+sstable_set::incremental_selector::selection
+sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
+    if (!_current_range_view || !_current_range_view->contains(pos, _cmp)) {
+        std::vector<versioned_sstable> _current_versioned_sstables;
+        std::tie(_current_range, _current_versioned_sstables, _current_next_position) = _impl->select(pos);
+        _current_sstables = boost::copy_range<std::vector<shared_sstable>>(_current_versioned_sstables
+                             | boost::adaptors::filtered([this] (versioned_sstable vsst) { return vsst->version_added <= _version && (!vsst->version_erased || vsst->version_erased > _version); })
+                             | boost::adaptors::transformed([] (versioned_sstable vsst) { return vsst->sst; }));
+        _current_range_view = _current_range->transform([] (const dht::ring_position& rp) { return dht::ring_position_view(rp); });
+    }
+    return {_current_sstables, _current_next_position};
+}
+
+sstable_set::incremental_selector
+sstable_set::make_incremental_selector(uint64_t version) const {
+    return incremental_selector(_impl->make_incremental_selector(), *_schema, version);
+}
+
+sstable_set::incremental_selector
+sstable_set::make_incremental_selector() const {
+    return incremental_selector(_impl->make_incremental_selector(), *_schema, _current_version);
+}
+
```

Comment:
This mixes code movement with changes. It's very hard to review. If you want to move code, move it either before or after this patch.

You can also send a separate pull request for code reorganization to get it out of the way.

Reply:
Ok, I'll add the code movement to a separate commit

Reply:
Was that done?

Reply:
It was done in the meantime in another patch that had already been merged: #7655